flagscale/train/hetero/parallel_context.py

```diff
@@ -508,28 +514,31 @@ def build_inter_mesh_process_groups(self, process_mesh1, process_mesh2):
             sp2 = cp2
         sp_overlapped_mapping = find_overlapped_mapping(sp1, sp2)
         dp_overlapped_mapping = find_overlapped_mapping(dp1, dp2)
-        dst_pp_dims = [0] 
+        src_pp_dims = [process_mesh1.get_parallel_size("pp") - 1]
+        dst_pp_dims = [0]
+        # i is cp, j is tp, k is dp, 
         for s in range(sp1):
             i, j = s % tp1, s // tp1
             for k in range(dp1):
-                src_coord = (i, j, k)
+                src_coord = [i, j, k, src_pp_dims[0]]
                 dst_sp_dims = [dim for dim, _, _ in sp_overlapped_mapping[s]]
                 dst_dp_dims = [dim for dim, _, _ in dp_overlapped_mapping[k]]
                 dst_coords = list(
                     itertools.product(dst_sp_dims, dst_dp_dims, dst_pp_dims)
                 )
-                src_rank = process_mesh1.logical_coords_to_ranks([src_coord])[0]
+                src_rank = process_mesh1.logical_coords_to_physical_ranks(
+                    [src_coord]
+                )[0]
                 for i, dst_coord in enumerate(dst_coords):
                     sp_dim, dp_dim, pp_dim = dst_coord
-                    dst_coord = (sp_dim % tp2, sp_dim // tp2, dp_dim, pp_dim)
+                    dst_coord = [sp_dim % tp2, sp_dim // tp2, dp_dim, pp_dim]
                     dst_rank = process_mesh2.logical_coords_to_physical_ranks(
                         [dst_coord]
                     )[0]
                     ranks = [src_rank, dst_rank]
                     timeout = max(process_mesh1._timeout, process_mesh2._timeout)
                     group = torch.distributed.new_group(ranks, timeout=timeout)
-                    if self._rank in [src_rank, dst_rank]:
```

Comment:
Each rank needs to store the `_inter_mesh_process_groups` information, so that during `_backtrack` we can judge whether two groups are inter-mesh connected.